src/devices/m_bus.c

```diff
@@ -63,7 +63,8 @@ static int m_bus_decode_3of6_buffer(uint8_t const *bits, unsigned bit_offset, ui
         uint8_t nibble_h = m_bus_decode_3of6(bitrow_get_byte(bits, n*12+bit_offset) >> 2);
         uint8_t nibble_l = m_bus_decode_3of6(bitrow_get_byte(bits, n*12+bit_offset+6) >> 2);
         if (nibble_h > 0xf || nibble_l > 0xf) {
-            return -1;
+            // return -1;  // fail at first 3of6 decoding error
+            nibble_l &= 0x0F;  // assume logical 0 nibble if 3of6 decoding error, let CRC fail decoding if necessary
```

Comment:
does this handle `nibble_h > 0xf`?
Shouldn't the goal be to return the number of successful decoded bytes instead of just an error (or hiding the error)?

Reply:
Ok, I see it does. We compose a partial byte filling the broken nibble with zero.
Should we end the decoding on first error or keep going?

Reply:
I thought it redundant to do `nibble_h &= 0x0F;` since the high part of that byte is getting immediately shifted away in `output[n] = (nibble_h << 4) | nibble_l;` (thus hiding any error in nibble_h as a side effect)

I agree it'd be more useful to have the return code be the number of decoded bytes found before the end or first error detected (instead of just `0` on success or `-1` on failure), though that change would require further changes.

Reply:
Are you thinking something like:

```
static int m_bus_decode_3of6_buffer(uint8_t const *bits, unsigned bit_offset, uint8_t* output, unsigned num_bytes)
{
    for (unsigned n=0; n<num_bytes; ++n) {
        uint8_t nibble_h = m_bus_decode_3of6(bitrow_get_byte(bits, n*12+bit_offset) >> 2);
        uint8_t nibble_l = m_bus_decode_3of6(bitrow_get_byte(bits, n*12+bit_offset+6) >> 2);
        if (nibble_h > 0xf || nibble_l > 0xf) {
            // return -1;  // fail at first 3of6 decoding error
            // nibble_l &= 0x0F;  // assume logical 0 nibble if 3of6 decoding error, let CRC fail decoding if necessary
            return n;  // return all successfully decoded bytes found until the first error
        }
        output[n] = (nibble_h << 4) | nibble_l;
    }
    return num_bytes;
}

```

It seems useful to me, but it looks like any positive result is then just categoriezed as success and then unused.  Should we change that behavior as well?

Reply:
I think a case could also be made to return the number of successful bytes until the first decode error (or even all bytes, regardless), still decoding all available bytes, zeroing any decoding errors, and let the caller decide if it's worth looking beyond the first decode error for other packets of data or try some kind of error correction.  Though perhaps the returns on this effort would be too diminished to be useful.

Reply:
Since there are no dynamic lengths involved I'd just decode everything (while zeroing errors) and return the number of good bytes. (`int bytes_ok = 0; … ++ ; return bytes_ok`)
Then let the caller decide, yes. As far as I understood these broken bytes are past the payload and CRC anyway, right? That's where my thinking came from, the "abort" is too harsh right now as we might not need all the bytes, so let the caller know how many good bytes we got and maybe that is enough.

Reply:
In theory we can use the crc to error correct. This 6 to 4 mapping could hint what nibble to error correct.

One thing that is in the specs but not in the code is that the number of bit transitions in the 6 bit code word is always greater then 1.

Im not saying we should add error correcting code now but maybe we could add some logic towards that.

In transmission theory the symbols can have an euclidean distance compared to each other. These 6 bit codes should have this property that if the code does not match any of the table codes if should be easy to map it to one code that is close in a euclidean way instead.

So the 6 to 4 decoder should output what it thinks the value should be and still somehow keep track of the indexes that might be incorrectly decoded and in need of error correction.

Reply:
Perhaps that kind of EC sounds like a job best left for rtl_wmbus?

What do you guys think of something like this:
```
static int m_bus_decode_3of6_buffer(uint8_t const *bits, unsigned bit_offset, uint8_t* output, unsigned num_bytes)
{
    int successful_contiguous_bytes = -1;
    for (unsigned n=0; n<num_bytes; ++n) {
        uint8_t nibble_h = m_bus_decode_3of6(bitrow_get_byte(bits, n*12+bit_offset) >> 2);
        uint8_t nibble_l = m_bus_decode_3of6(bitrow_get_byte(bits, n*12+bit_offset+6) >> 2);
        if (nibble_h > 0xf || nibble_l > 0xf) {
            // return -1;  // fail at first 3of6 decoding error
            nibble_l &= 0x0F;  // assume logical 0 nibble if 3of6 decoding error, let CRC fail decoding if necessary
            successful_contiguous_bytes > -1  || successful_contiguous_bytes = n;  // return count found until the first error
        }
        output[n] = (nibble_h << 4) | nibble_l;
    }
    successful_contiguous_bytes > -1  || successful_contiguous_bytes = num_bytes;  // if all data decoded successfully
    return successful_contiguous_bytes;
}

```

Reply:
Fine by me.

Reply:
I don't think 3of6 lends itself well to EC, e.g. assume a single bit error in `001111`, would this be `001101`, `001110`, or `001011`?

Reply:
Well the CRC will probably only pass with one of these values. But the complexity of doing EC 2 steps back will get quite complicated quickly. In practice all 6-bit codes that are invalid has a valid value that is most likely to be the correct one. So if one had lots of time to invest into this all kinds of clever heuristics could be developed.

Reply:
I do think keeping track of nibble indices that 6to4 decodes improperly and then brute force trying all 4-bit combinations and calculating the CRC over that would be just as effective at doing EC compared to some clever heuristics in the 6to4 decoder.

Reply:
Do we need to implement EC as part of this change?